### DISCUSSION TITLE
使用工程移植tinyusb

### DISCUSSION THREAD

**Discussion by miaoxiaohu** [GorgonMeducer/Pico_Template#3] [Q&A]
你好，最近使用你的工程，尝试移植tinyusb，编译没问题，但是插上usb口不回应主机数据。
就是想问下，适用这个mdk工程移植tinyusb有问题吗？

**Comment by GorgonMeducer** [comment]
我没试过TinyUSB的移植。但有一点，切记不可以在使用CMSIS-DAP进行调试的情况下进行，因为pico-debug 这个固件已经占用USB了。
如果是调试USB，可能要搞个USB逻辑分析仪，看看问题出在哪里了。

**Reply by miaoxiaohu** [reply]
好的，我再用逻辑分析仪查查问题，谢谢回复，这个工程模块很好用。

**Reply by GorgonMeducer** [reply]
有几个思路给你参考下：
1、armclang 的优化比 gcc强很多，建议检查下tinyUSB有没有什么对时序很敏感的代码。如果确实发现了，除了加 volatile 以外，推荐加 __NOP()（包含 cmsis_compiler.h以后可以使用）
2、尝试关闭 armclang 的 link-time-optimisation 以后问题是否有所改善（倒不是lto有什么bug，而是他会让很多时序敏感的代码变的更为敏感）
3、尝试用 -O0 检查逻辑。如果可以解决问题，那么就用MDK工程管理器里可以给group或者单个源文件设置优化等级的方式来通过缩小范围的方法找到具体是哪个源文件对时序敏感了。注意，如果要关闭 LTO，除了不勾选Link-time-optimisation以外，在Misc Control里加 -fno-lto 更为有效。
